tests/query_params_test.go

```diff
@@ -0,0 +1,216 @@
+		}
+		if vegetableParamValues[0] != valueOne || vegetableParamValues[1] != valueTwo {
+			t.Errorf("queryParams values: expected "+valueOne+" and "+valueTwo+", got %s and %s", vegetableParamValues[0], vegetableParamValues[1])
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+func TestQueryParametersSameKeyDifferentCasing(t *testing.T) {
+
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+
+		if len(queryParams) != 2 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 2)
+		}
+
+		if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
+			t.Errorf("Expected one value for query parameter vegetables and one for Vegetables, got %d and %d", len(queryParams["vegetable"]), len(queryParams["Vegetable"]))
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+}
+
+// TestQueryParametersValidUnicode ensures keys and values that contain non-ASCII characters are parsed correctly
+func TestQueryParametersValidUnicode(t *testing.T) {
+	value := "ăȚâȘî"
+	req := []byte("GET /?vegetable=" + value + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if valueReceived := req.URL.Query()["vegetable"][0]; valueReceived != value {
+			t.Errorf("queryParams values: got %s, want %s", valueReceived, value)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+	key := "ăȚâȘî"
+	req = []byte("GET /?" + key + "=vegetable HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if listLen := len(req.URL.Query()[key]); listLen != 1 {
+			t.Errorf("len(queryParamsKey): got %d, want 1 value for %s", listLen, key)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+// Tests whether passing invalid Unicode will result  in a 400 Bad Request response
+func TestQueryParametersInvalidUnicodes(t *testing.T) {
+	key := "\x0F"
+	req := []byte("GET /?" + key + "=tomato&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
```

Comment:
Consider dumping the request here to give the developer who breaks the test as much information as possible.

https://golang.org/pkg/net/http/httputil/#DumpRequest

Reply:
I'd opt against it because we would depend on one extra layer and the request dumped is an approximation according to the documentation. We would like to have all the checks in the handler to be as close as possible to the real usage.